### ISSUE TITLE
Bug in opnd_is_memory_ref

### ISSUE THREAD

**Issue by johnfxgalea** [DynamoRIO/dynamorio#4260]
opnd_is_memory_ref() should check for absolute addr not only for X86_64 but for X64 overall.

https://github.com/DynamoRIO/dynamorio/blob/4f9b46023f05c1cfb756e0c61085715a0126b657/core/arch/opnd_shared.c#L1002

opnd_create_far_abs_addr() creates relative opnd but still assings the ABS_ADDR_kind flag to the opnd. Hence, if this function is called on aarch64, opnd_is_memory_ref() will incorrectly return false.

**Comment by johnfxgalea** [comment]
@derekbruening Do you agree with my reading please?

**Comment by derekbruening** [comment]
I think that opnd_create*_abs_addr() will fail to encode with any opcode on AArch64 today.

The docs say that it will be converted to PC-relative (that's what you mean by "creates relative opnd" right?), but only x86 seems to have code to do that today.

Maybe let's expand this issue to cover all of these problems?  Action items are:

1) Decide whether to support opnd_create_abs_addr() for AArch64.  The problem is that using PC-relative requires a change in opcode: LDR or LDRSW (are there even any PC-relative store instructions?), so unlike x86 (or ARM IIRC) it's not an operand-local decision.  Maybe we should disallow (not sure whether ifdef in header or just document and fail in impl is better) and provide an XINST_CREATE_pcrel_load() or sthg.

2) Update opnd_create_abs_addr() docs: it talks about `rax` without a mention of that being x86-only.

@AssadHashmi WDYT?

**Comment by johnfxgalea** [comment]
> The docs say that it will be converted to PC-relative (that's what you mean by "creates relative opnd" right?), but only x86 seems to have code to do that today.

Yes

**Comment by AssadHashmi** [comment]
> are there even any PC-relative store instructions?

No, the PC can't be explicitly referenced as the base, index or transfer register in any load/store instructions. As you say, only the `LDR <Xt>, <label>` and `LDRSW <Xt>, <label>` instructions implicitly read the PC and calculate address from the PC's value and an immediate offset.

> Maybe we should disallow (not sure whether ifdef in header or just document and fail in impl is better) 

IMO we should disallow with an ifdef.

> and provide an XINST_CREATE_pcrel_load() or sthg

Do you mean use XINST_CREATE_pcrel_load() for `LDR <Xt>, <label>` and `LDRSW <Xt>, <label>` ?

**Comment by derekbruening** [comment]
> > and provide an XINST_CREATE_pcrel_load() or sthg
> 
> Do you mean use XINST_CREATE_pcrel_load() for `LDR <Xt>, <label>` and `LDRSW <Xt>, <label>` ?

Right.  Though I just looked them up and they both have only +-1MB reach, so they can't be used to reach .data in a client library from the code cache, so maybe they would almost never be used.

This also brings up opnd_create_rel_addr(), which also seems less useful for AArch64 (I guess ARM too which also has short reach).  I suppose custom generated code could embed data, but typical uses of opnd_create_abs_addr() and opnd_create_rel_addr() that come to mind are reaching client libraries.

**Comment by johnfxgalea** [comment]
I'd go for an ifdef too

**Comment by AssadHashmi** [comment]
> Right. Though I just looked them up and they both have only +-1MB reach, so they can't be used to reach .data in a client library from the code cache, so maybe they would almost never be used.

Good point! Do we have documentation of DynamoRIO internals which describes the memory layout, especially in relation to managing the code cache? Or can you point to the relevant areas of code?

> This also brings up opnd_create_rel_addr(), which also seems less useful for AArch64 (I guess ARM too which also has short reach). I suppose custom generated code could embed data, but typical uses of opnd_create_abs_addr() and opnd_create_rel_addr() that come to mind are reaching client libraries.

Lack of these addressing features has not impeded our work...yet. Perhaps for now we should ifdef them out but keep this issue open as a placeholder?

**Comment by derekbruening** [comment]
> > Right. Though I just looked them up and they both have only +-1MB reach, so they can't be used to reach .data in a client library from the code cache, so maybe they would almost never be used.
> 
> Good point! Do we have documentation of DynamoRIO internals which describes the memory layout, especially in relation to managing the code cache? Or can you point to the relevant areas of code?

Yes, we put some effort into keeping certain structures within x86 PC-relative reach of each other, providing options for allocating heap either "reachable" or not, etc.  The currently-x86-centric docs are here, talking about reaching client libraries, reaching heap, and absolute and pc-rel operands: https://dynamorio.org/dynamorio_docs/using.html#sec_64bit_reach

What is the longest reach of a PC-relative branch on AArch64?  If that is also <<2GB, these docs should probably be qualified when they talk about reaching to be referring only to x86 PC-relative branches and data references.  The docs there about absolute (and pc-rel) operands should also be updated alongside changes to disable or qualify the operand code.

> 
> > This also brings up opnd_create_rel_addr(), which also seems less useful for AArch64 (I guess ARM too which also has short reach). I suppose custom generated code could embed data, but typical uses of opnd_create_abs_addr() and opnd_create_rel_addr() that come to mind are reaching client libraries.
> 
> Lack of these addressing features has not impeded our work...yet. Perhaps for now we should ifdef them out but keep this issue open as a placeholder?

It just means you need an extra instruction (and register in some cases) to set up the address.  E.g., recently it showed up in drbbdup where it loads form memory to a register and assumes it can do that in one instruction: https://github.com/DynamoRIO/dynamorio/blob/master/ext/drbbdup/drbbdup.c#L750
It is in several other places in other clients..I can try to find them if you like.
We already have some helpers functions in place for this "might need 2 instructions" for some kinds of things: instrlist_insert_mov_immed_ptrsz(), instrlist_insert_push_immed_ptrsz(), and the two instr_addr() variants.  Maybe we could have a similar one for loading something from an absolute address?  @johnfxgalea would that help drbbdup?

**Comment by derekbruening** [comment]
(For clean calls: on x86 they reach client libraries b/c of the layout, but on aarch they need to be indirect with extra instructions to set up the address in a register -- but IIRC the code for that is all in place now and it done auto-magically, including on x86 if a client targets some far-away callee.)

**Comment by johnfxgalea** [comment]
Assuming we have locked loads, drbbdup doesn't care how many instructions it takes to load from an abs pointer.

**Comment by AssadHashmi** [comment]
> What is the longest reach of a PC-relative branch on AArch64? If that is also <<2GB, these docs 

AFAIK the longest PC-relative branch is +/- 128MB

**Comment by johnfxgalea** [comment]
So what are we going to do please? ifdef out + helper function?

**Comment by AssadHashmi** [comment]
> So what are we going to do please? ifdef out + helper function?

That would be my vote, at least we  can do the ifdef for now.

**Comment by derekbruening** [comment]
I assume that whatever we do for AArch64 we would do for 32-bit ARM as well wrt abs addrs, since it also always requires a base register.

My hesitation in ifdef-ing out is: for cross-platform code it could be nice to have one of these, the abs or rel operand, work on all platforms.  E.g., you could imagine this:
```
XINST_CREATE_load(dc, opnd_create_reg(myreg), opnd_create_abs_addr(nearby_addr));
```

If we don't ifdef out abs addrs on AArch64 and ARM, we could instead treat them just like relative addrs, and allow that line of code to then work on all 4 architectures.

Or we could instead make the user use opnd_create_rel_addr there, and augment today's x86 conversion of rel to abs to also convert to any baseless and not just the eax immed-addr forms.

Or perhaps make the user always use a higher-level construct as I was suggesting up above which picks the opcode and may use multiple instructions.

I suspect the user almost never cares about the distinction between opnd_create_abs_addr() and opnd_create_rel_addr(): the user just wants to load or store to an address without spilling a base register.  This is why the two will convert to each other in some cases on x86, to make it easier to write code that works on 32-bit x86 and 64-bit x86.  Maybe we should never have created two different types?

**Comment by derekbruening** [comment]
My logic for preferring the opnd_create_abs_addr() is that having "absolute address" turn into PC-relative under the covers seems perfectly logical to me, while having "relative address" turn into absolute seems a little weird.

But, given the limited reaches on ARM and AArch64, maybe disallowing "absolute address" and forcing the user to think about it being relative is good.

To get a visual feel, we're looking at:

```C
XINST_CREATE_load(dc, opnd_create_reg(dest_reg), opnd_create_abs_addr(literal_addr));
vs
XINST_CREATE_load(dc, opnd_create_reg(dest_reg), opnd_create_rel_addr(literal_addr));
vs
instrlist_insert_load_from_literal(dc, ilist, where, literal_addr, dest_reg, scratch_reg, NULL, NULL)
```

**Comment by AssadHashmi** [comment]
I think I understand what you mean and your intent.
We'll need to think about this. Will get back to you next week if that's ok?

**Comment by derekbruening** [comment]
Sure, there's no rush.

Re: the drbbdup usage: that wants a load-acquire which is a mostly separate discussion (maybe for #4215), though there is some overlap and some parallel issues: trying to use the x86 LOCK() allowing for granular use around various loads and stores, versus a higher-level version which then has to take in enough params to support various data sizes or other variations, or an extra param to instrlist_insert_load_from_literal about the memory model to enforce.

Re: should we get rid of opnd_create_rel_addr(): with automated re-relativization (recently extended to branches on x86: #731,#3271; still missing on AArchXX: #4016), a PC-relative encoded operand is automatically updated to keep pointing at the same absolute addr if re-encoded elsewhere.  It seems like a relative addr operand is a logically absolute addr operand with an encoding hint, sort of like a base-disp x86 operand with an encoding hint to please encode with a displacement even if it's 0 (b/c the user wants to patch in a different displacement later).
